# Transformers in Single-Cell Omics

This repository accompanies [Transformers in Single-Cell Omics:
A Review and New Perspectives](https://rdcu.be/dQuVW). Please refer to the manuscript for the details.

We provide a curated list of single-cell transformers and their evaluation results. We skip models that work only on bulk or images of slides data and those where transformers are used only as a part of the model. Models focusing on sequential data, such as DNA or protein sequences are omitted too. New entries are added at the top of the corresponding table.

We welcome contributions to this repository. Please open a pull request or an issue if you want to add or edit an entry.

## Single-cell transformers

| Model                    | Paper                                                                                                                                                              | Code                                                                               | Omic Modalities                                                | Pre-training Dataset                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Input Embedding                                                                                                                                                                                        | Architecture                                                         | SSL Tasks                                                                                                                   | Supervised Tasks                                                                                                                                               | Zero-shot Tasks                                                                                                              |
|--------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------|----------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------|
| Precious3GPT             | üìù[Galkin et al. 2024](https://doi.org/10.1101/2024.07.25.605062)                                                                                                  | [Partial üîçÔ∏èü§ó](https://huggingface.co/insilicomedicine/precious3-gpt-multi-modal) | Bulk/scRNA-seq, DNAm, proteomics, natural language annotations | Omics data with KG and text embeddings, Closed source                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | ?                                                                                                                                                                                                      | Decoder-only LLaMA-like transformer model with modality mapper units | Emulation of chemical response, cross-species/tissue/omics transference, emulation of clinical conditions                   | Age prediction, gene classification                                                                                                                            | DEG prediction, Indication discovery                                                                                         |
| LangCell                 | üìÑ(ICML)[Zhao et al. 2024](https://arxiv.org/abs/2405.06708)                                                                                                       | [üõ†Ô∏èGithub](https://github.com/PharMolix/LangCell)                                 | scRNA-seq, natural language                                    | 27M / cross-tissue, human ([CELLxGENE](https://cellxgene.cziscience.com/))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | Ordering: rank-based, natural language cell description                                                                                                                                                | Other: two encoders (cell and text)                                  | MLM with CE loss, intra- and inter-modal contrastive loss, cell-text matching with CE loss                                  | Cell type annotation, pathway identification                                                                                                                   | Novel cell type identification, NSCLC subtype classification, batch integration, cell clustering                             | 
| ScRAT                    | üìÑ(Bioinformatics)[Mao et al. 2024](https://academic.oup.com/bioinformatics/article/40/2/btae067/7613064)                                                          | [üõ†Ô∏èGitHub](https://github.com/yuzhenmao/ScRAT)                                    | scRNA-seq                                                      | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Cells as tokens                                                                                                                                                                                        | Encoder                                                              | None                                                                                                                        | Phenotype prediction: aggregated per sample cell embeddings are used to predict sample label (e.g., health condition)                                          | None                                                                                                                         | 
| scPRINT                  | üìù[Kalfon et al. 2024](https://www.biorxiv.org/content/10.1101/2024.07.29.605556v1)                                                                                | [üõ†Ô∏èGithub](https://github.com/cantinilab/scPRINT)                                 | scRNA-seq                                                      | 50M / cross-tissue, cross-species ([CELLxGENE](https://cellxgene.cziscience.com/))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | Other: [ESM-2](https://www.science.org/doi/10.1126/science.ade2574) based gene embeddings. Gene embeddings are randomly sampled and order determined by position on chromosomes                        | Encoder                                                              | Multi task Pre-training: Denoising, Botleneck learning (+ many additional losses available)                                 | Cell label prediction (these supervised tasks are part of the pre-training)                                                                                    | Read depth enhancement, gene expression imputation, Batch Integration, Cell Clustering, Cell Label Prediction, GRN inference |
| scMulan                  | üìÑ(RECOMB)[Bian et al. 2024](https://link.springer.com/chapter/10.1007/978-1-0716-3989-4_57)                                                                       | [üîçGithub](https://github.com/SuperBianC/scMulan/tree/main)                        | scRNA-seq                                                      | 10M / cross-tissue, human ([hECA](https://www.sciencedirect.com/science/article/pii/S2589004222005892))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | Not specified                                                                                                                                                                                          | Decoder                                                              | Conditional cell generation                                                                                                 | cell type annotation, cell metadata annotation (both also used in training)                                                                                    | Batch integration                                                                                                            |
| BioFormers               | üìù[Belgadi and Li et al. 2023](https://www.biorxiv.org/content/10.1101/2023.11.29.569320v1.full.pdf)                                                               | None                                                                               | scRNA-seq                                                      | 8K / single tissue, human ([PBMC](https://docs.scvi-tools.org/en/stable/api/reference/scvi.data.pbmc_dataset.html), [Adamson et al. 2016](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE90546))                                                                                                                                                                                                                                                                                                                                                                                                                            | Value categorization: value binning                                                                                                                                                                    | Encoder                                                              | MLM with CE loss                                                                                                            | None                                                                                                                                                           | Cell clustering, gene expression imputation, genetic perturbation effect prediction, GRN inference                           |
| Geneformer               | üìÑ(Nature)[Theodoris et al. 2023](https://www.nature.com/articles/s41586-023-06139-9)                                                                              | [üõ†ü§ó](https://huggingface.co/ctheodoris/Geneformer)                               | scRNA-seq                                                      | 36M / cross-tissue, human ([Genecorpus](https://huggingface.co/datasets/ctheodoris/Genecorpus-30M))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | Ordering: rank-based                                                                                                                                                                                   | Encoder                                                              | MLM with CE loss, gene ID prediction                                                                                        | Gene function prediction, cell annotation                                                                                                                      | Cell clustering, GRN inference                                                                                               |
| Universal Cell Embedding | üìù[Rosen et al. 2023](https://www.biorxiv.org/content/10.1101/2023.11.28.568918v1.full.pdf)                                                                        | [üîçGithub](https://github.com/snap-stanford/uce)                                   | scRNA-seq                                                      | 36M / cross-tissue, cross-species ([CELLxGENE](https://cellxgene.cziscience.com/) and [other](https://www.biorxiv.org/content/biorxiv/early/2023/11/29/2023.11.28.568918/DC3/embed/media-3.xlsx?download=true))                                                                                                                                                                                                                                                                                                                                                                                                                    | Other: [ESM-2](https://www.science.org/doi/10.1126/science.ade2574) based gene embeddings. Gene embeddings are sampled according to expression levels and order determined by position on chromosomes. | Encoder                                                              | Modified MLM, binary CE loss predicting whether a gene is expressed or not. Uses CLS embedding instead of token-embeddings. | Cell annotation                                                                                                                                                | Cell clustering, cross-species integration                                                                                   |
| scGPT                    | üìÑ(Nature Meth)[Cui et al. 2024](https://www.nature.com/articles/s41592-024-02201-0)                                                                               | [üîçGitHub](https://github.com/bowang-lab/scGPT)                                    | scRNA-seq, scATAC-seq, CITE-seq, Spatial transcriptomics       | 33M / cross-tissue, human, non-disease ([CELLxGENE](https://cellxgene.cziscience.com/))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | Value categorization: value binning                                                                                                                                                                    | Other: attention masking in encoder                                  | Iterative MLM variant with MSE loss, cell token expression prediction, gene expression prediction                           | Cell type annotation, genetic perturbation effect prediction, reverse perturbation prediction, cell clustering, multimodal embedding, gene function prediction | Cell clustering, GRN inference, simulation, gene expression imputation                                                       |
| TOSICA                   | üìÑ(Nature Comms)[Chen et al. 2023](https://www.nature.com/articles/s41467-023-35923-4)                                                                             | [üõ†Ô∏èGitHub](https://github.com/JackieHanLab/TOSICA)                                | scRNA-seq                                                      | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Value projection                                                                                                                                                                                       | Encoder                                                              | None                                                                                                                        | Cell type annotation                                                                                                                                           | None                                                                                                                         |
| scMoFormer               | üìÑ(ACM)[Tang et al. 2023](https://dl.acm.org/doi/10.1145/3583780.3615061)                                                                                          | [üõ†Ô∏èGitHub](https://github.com/OmicsML/scMoFormer)                                 | scRNA-seq, scATAC-seq, CITE-seq                                | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Other, SVD-based                                                                                                                                                                                       | Encoder and graph transformers                                       | None                                                                                                                        | Cross-modality prediction                                                                                                                                      | None                                                                                                                         |
| tGPT                     | üìÑ(Cell iScience)[Shen et al. 2023](https://www.cell.com/iscience/pdf/S2589-0042(23)00613-2.pdf)                                                                   | [üõ†GitHubÔ∏è](https://github.com/deeplearningplus/tGPT)                              | scRNA-seq                                                      | 22M / cross-tissue, cross-species, disease and non-disease, organoids ([list](https://www.cell.com/cms/10.1016/j.isci.2023.106536/attachment/1e95114b-5ea0-4596-afd8-0cb04bae0f6d/mmc2))                                                                                                                                                                                                                                                                                                                                                                                                                                           | Ordering                                                                                                                                                                                               | Decoder                                                              | NTP with CE loss, gene ID prediction                                                                                        | None                                                                                                                                                           | Cell clustering, trajectory inference                                                                                        |
| SpaFormer                | üìù[Wen et al. 2023](https://arxiv.org/abs/2302.03038)                                                                                                              | [üõ†Ô∏èGitHub](https://github.com/wehos/CellT)                                        | Spatial transcriptomics                                        | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Cells as tokens, value projection                                                                                                                                                                      | Encoder                                                              | Modified MLM with MSE loss, gene expression prediction                                                                      | Gene expression imputation                                                                                                                                     | Cell clustering                                                                                                              |
| scFoundation             | üìÑ(Nature Meth)[Hao et al. 2024](https://www.nature.com/articles/s41592-024-02305-7) and üìÑ(NeurIPS)[Gong et al. 2023](https://openreview.net/forum?id=gdwcoBCMVi) | [üîçGitHub](https://github.com/biomap-research/scFoundation)                        | scRNA-seq                                                      | 50M / cross-tissue, human, disease and non-disease ([GEO](https://www.ncbi.nlm.nih.gov/geo/), [Single Cell Portal](https://singlecell.broadinstitute.org/single_cell), [HCA](https://data.humancellatlas.org/), [EMBL-EBI](https://www.ebi.ac.uk/))                                                                                                                                                                                                                                                                                                                                                                                | Value projection                                                                                                                                                                                       | Other: two encoders                                                  | Modified MLM with MSE loss, gene expression prediction                                                                      | Drug response prediction, genetic perturbation effect prediction                                                                                               | Read depth enhancement, cell clustering                                                                                      |
| CellLM                   | üìù[Zhao et al. 2023](https://arxiv.org/abs/2306.04371)                                                                                                             | [üîçGitHub](https://github.com/PharMolix/OpenBioMed)                                | scRNA-seq                                                      | 1.8M / cross-tissue, human, disease and non-disease ([PanglaoDB](https://panglaodb.se/), [CancerSCEM](https://ngdc.cncb.ac.cn/cancerscem/))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | Value categorization                                                                                                                                                                                   | Encoder                                                              | Contrastive loss, MLM with CE loss                                                                                          | Non-disease vs cancer prediction, cell type annotation, drug response prediction                                                                               | None                                                                                                                         |
| scCLIP                   | üìù[Xiong et al. 2023](https://openreview.net/forum?id=KMtM5ZHxct&referrer=%5Bthe%20profile%20of%20Tianlong%20Chen%5D(%2Fprofile%3Fid%3D~Tianlong_Chen1))           | [üõ†Ô∏èGitHub](https://anonymous.4open.science/r/scCLIP-61F6/README.md)               | scRNA-seq, scATAC-seq                                          | 377k / cross-tissue, human fetal ([ATAC](https://www.science.org/doi/10.1126/science.aba7612), [RNA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7780123/))                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | Value projection                                                                                                                                                                                       | Encoder                                                              | Contrastive loss, CE matching modalities                                                                                    | None                                                                                                                                                           | Multimodal embedding                                                                                                         |
| GeneCompass              | üìù[Yang et al. 2023](https://www.biorxiv.org/content/10.1101/2023.09.26.559542v1)                                                                                  | [Partial üõ† GitHub](https://github.com/xCompass-AI/GeneCompass)                    | scRNA-seq                                                      | 126M / cross-tissue, human and mouse, disease and non-disease ([GEO](https://www.ncbi.nlm.nih.gov/geo/), [SRA](https://www.ncbi.nlm.nih.gov/sra), [CELLxGENE](https://cellxgene.cziscience.com/), [GSA](https://ngdc.cncb.ac.cn/gsa/), [Single Cell Portal](https://singlecell.broadinstitute.org/single_cell), [HCA](https://data.humancellatlas.org/), [EMBL-EBI](https://www.ebi.ac.uk/), [3CA](https://www.weizmann.ac.il/sites/3CA/), [Cell BLAST](https://cblast.gao-lab.org/), [TEDD](https://TEDD.obg.cuhk.edu.hk/),  and [other](https://www.biorxiv.org/content/10.1101/2023.09.26.559542v1.full))                       | ?                                                                                                                                                                                                      | Other: two encoders                                                  | MLM with CE and MSE loss, gene ID and expression prediction                                                                 | Cell type annotation, drug response prediction, gene function prediction                                                                                       | Cross-species integration, genetic perturbation effect prediction, GRN inference                                             |
| CellPLM                  | üìÑ(ICLR)[Wen et al. 2024](https://openreview.net/forum?id=BKXvPDekud)                                                                                              | [Partial üîçGitHub](https://github.com/OmicsML/CellPLM)                             | scRNA-seq, Spatial transcriptomics                             | 11M / cross-tissue, human, disease and non-disease ([HTCA](https://humantumoratlas.org/), [HCA](https://data.humancellatlas.org/), [GEO](https://www.ncbi.nlm.nih.gov/geo/))                                                                                                                                                                                                                                                                                                                                                                                                                                                       | Cells as tokens, value projection                                                                                                                                                                      | Encoder                                                              | Modified MLM with MSE loss and KL losses, gene expression prediction                                                        | Gene expression imputation, cell type annotation, genetic perturbation effect prediction                                                                       | Cell clustering, scRNA-seq denoising                                                                                         |
| scMAE                    | üìù[Kim et al. 2023](https://openreview.net/pdf?id=2mq6uezuGj)                                                                                                      | None                                                                               | single-cell flow cytometry                                     | 6.5M / human, disease and non-disease (source?)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | Other, concatenation of values with learnable protein embeddings                                                                                                                                       | Other: two encoders                                                  | MLM with MSE loss, protein expression prediction                                                                            | Cell type annotation, protein expression imputation                                                                                                            | None                                                                                                                         |
| CAN/CGRAN                | üìÑ(Frontiers)[Wang et al. 2023](https://ebooks.iospress.nl/volumearticle/64489)                                                                                    | None                                                                               | scRNA-seq                                                      | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Value projection                                                                                                                                                                                       | Encoder                                                              | None                                                                                                                        | Cell type annotation                                                                                                                                           | None                                                                                                                         |
| scTranslator             | üìù[Liu et al. 2023](https://www.biorxiv.org/content/10.1101/2023.07.04.547619v2.full)                                                                              | [üîçÔ∏èGitHub](https://github.com/TencentAILabHealthcare/sctranslator)                | scRNA-seq, CITE-seq                                            | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Value projection                                                                                                                                                                                       | Other: two encoders                                                  | None                                                                                                                        | Cross-modality prediction                                                                                                                                      | (After cross-modality prediction training) GRN inference, cell clustering                                                    |
| scTransSort              | üìÑ(MDPI)[Jiao et al. 2023](https://www.mdpi.com/2218-273X/13/4/611)                                                                                                | [üõ†Ô∏èGitHub](https://github.com/jiaojiao-123/scTransSort)                           | scRNA-seq                                                      | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Value projection                                                                                                                                                                                       | Encoder                                                              | None                                                                                                                        | Cell type annotation                                                                                                                                           | None                                                                                                                         |
| STGRNS                   | üìÑ(OUP)[Xu et al. 2023](https://academic.oup.com/bioinformatics/article/39/4/btad165/7099621)                                                                      | [üõ†Ô∏èGitHub](https://github.com/zhanglab-wbgcas/STGRNS)                             | scRNA-seq                                                      | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Other                                                                                                                                                                                                  | Encoder                                                              | None                                                                                                                        | GRN inference                                                                                                                                                  | None                                                                                                                         |
| CIForm                   | üìÑ(OUP)[Xu et al. 2023](https://academic.oup.com/bib/article-abstract/24/4/bbad195/7169137?redirectedFrom=fulltext)                                                | [üõ†Ô∏èGitHub](https://github.com/zhanglab-wbgcas/CIForm)                             | scRNA-seq                                                      | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Value projection                                                                                                                                                                                       | Encoder                                                              | None                                                                                                                        | Cell type annotation                                                                                                                                           | None                                                                                                                         |
| scFormer                 | üìù[Cui et al. 2023](https://openreview.net/forum?id=7hdmA0qtr5)                                                                                                    | [Incomplete Ô∏èGitHub](https://github.com/bowang-lab/scFormer)                       | scRNA-seq                                                      | Task specific                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | Value categorization: value binning                                                                                                                                                                    | Encoder                                                              | Modified MLM with CE, cell token expression prediction, contrastive loss with cosine similarity, gene expression prediction | Cell type annotation, genetic perturbation effect prediction                                                                                                   | Cell clustering                                                                                                              |
| Exceiver                 | üìù[Connell et al. 2022](https://openreview.net/forum?id=XxRuCIgq2LX)                                                                                               | [üõ†Ô∏èGitHub](https://github.com/keiserlab/exceiver)                                 | scRNA-seq                                                      | 0.5M / cross-tissue, human ([Tabula Sapiens](https://figshare.com/ndownloader/files/34702114))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | Other: value scaled embeddings                                                                                                                                                                         | Encoder                                                              | Modified MLM with MSE, gene expression prediction                                                                           | Cell type annotation, drug response prediction                                                                                                                 | Cell clustering                                                                                                              |
| TransCluster             | üìÑ(Frontiers)[Song et al. 2022](https://www.frontiersin.org/articles/10.3389/fgene.2022.1038919/full)                                                              | [üõ†Ô∏èGitHub](https://github.com/Danica123/TransCluster)                             | scRNA-seq                                                      | None                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Value projection with LDA                                                                                                                                                                              | Encoder                                                              | None                                                                                                                        | Cell type annotation                                                                                                                                           | None                                                                                                                         |
| scBERT                   | üìÑ(Nature MI)[Yang et al. 2022](https://www.nature.com/articles/s42256-022-00534-z)                                                                                | [üîçGitHub](https://github.com/TencentAILabHealthcare/scBERT)                       | scRNA-seq                                                      | 1M / cross-tissue, human ([PanglaoDB](https://panglaodb.se/))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | Value categorization, binning                                                                                                                                                                          | Encoder                                                              | MLM with CE loss, gene expression prediction                                                                                | Cell type annotation, unseen cell type detection                                                                                                               | None                                                                                                                         |
| iSEEEK                   | üìÑ(OUP)[Shen et al. 2022](https://academic.oup.com/bib/article/23/2/bbab573/6511497?login=false)                                                                   | [üîçGithub](https://github.com/lixiangchun/iSEEEK) (dataset not public)             | scRNA-seq                                                      | 11.9M / cross-tissue, cross-species ([list](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/23/2/10.1093_bib_bbab573/1/supplementary_table_1_data_source_information_bbab573.xlsx?Expires=1710130756&Signature=EKqLNLeqpYl320~rRpsK79fA4RC1KvQcWeYqpGALwIQ2SK-IAqZmyXnU-tuYG4xzZqeOhjKc2oyhqBRkO8xJBGO~EMja5KhbNP8PS4nV1eFCrZW6GvbluqyqCP9v-z2ExYhSo3f4jisWi9irRL0y7fxIPk3dPV4f3NACrRjDinqpUZ2eJTKOWwQ8GTbL3dySxPNJa8XQ0y2lGv0lBO~KMG-PJROYDi-PQyHHkTz5AbAnhpD5jnK1BSdxNg8oFuIErS7U0ej8V86eA3AqZyFdI4RZtq-iPTkAEFSQROhQdv1aTh3Cj5EuIZvVgmivr~qF571C9m4IHnI0bEsrhI4f3w__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA)) | Ordering: rank-based                                                                                                                                                                                   | Encoder                                                              | MLM with CE loss                                                                                                            | Marker gene classification                                                                                                                                     | Cell clustering, pseudotime analysis, GRN inference                                                                          |
| Multitask learning       | üìù[Pang et al. 2020](https://www.biorxiv.org/content/10.1101/2020.02.05.935239v2)                                                                                  | None                                                                               | scRNA-seq                                                      | 160k / brain, mouse ([MBA](http://mousebrain.org/))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | Value projection                                                                                                                                                                                       | Other: autoencoder with two transformer encoders (?)                 | Modified MLM with MSE loss, gene expression prediction                                                                      | None                                                                                                                                                           | Cell clustering                                                                                                              |

## Transformer LLMs for single-cell

| Model          | Paper                                                                                                                                                                     | Code                                                        | Omic Modalities                    | Pre-training Dataset                                                                                                                                                                                                                                                                                                                                       | Input Embedding                                                                                                                  | Architecture                                                                                                        | SSL Tasks                                                           | Supervised Tasks                                                       | Zero-shot Tasks                                                                                          |
|----------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------|------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------|------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------|
| CELLama        | üìù[Choi et al. 2024](https://www.biorxiv.org/content/10.1101/2024.05.08.593094v1.full#ref-16)                                                                             | [üõ†GitHub](https://github.com/portrai-io/CELLama)           | scRNA-seq, Spatial transcriptomics | Natural Language [SBERT](https://fq.pkwyx.com/default/https/aclanthology.org/D19-1410.pdf)                                                                                                                                                                                                                                                                 | Other: Ordering with embedding of the natural language representation, additional cell annotations are added in natural language | Siamese encoders ([SBERT](https://fq.pkwyx.com/default/https/aclanthology.org/D19-1410.pdf))                        | Contrastive loss                                                    | Cell type annotation                                                   | Cell type annotation, niche cell type featuring                                                          |
| CellWhisperer  | üìù[Schaefer et al. 2024](https://openreview.net/forum?id=yWiZaE4k3K)                                                                                                      | Soon                                                        | Bulk/scRNA-seq                     | Transcriptome data paired with natural language annotations                                                                                                                                                                                                                                                                                                | Geneformer- and BioBERT-based embedding models (contrastively fine-tuned)                                                        | Multimodal contrastive training of embedding models (CLIP) and transcriptome instruction fine-tuning of LLM (LLaVA) | None                                                                | Transcriptome-aware question-answering                                 | Reference-free cell property prediction (cell types & states, disease states, organ of cell origin, ...) |
| scInterpreter  | üìù[Li et al. 2024](https://arxiv.org/abs/2402.12405)                                                                                                                      | None                                                        | scRNA-seq                          | Natural Language [GPT-3.5](https://api.openai.com/) and [Llama-13b](https://arxiv.org/abs/2302.13971)                                                                                                                                                                                                                                                      | Other: Ordering with embedding of the natural language representation                                                            | Decoder, [GPT-3.5](https://api.openai.com/) and [Llama-13b](https://arxiv.org/abs/2302.13971)                       | NTP with CE loss and instruction finetuning (GPT-3.5 closed-source) | None                                                                   | Cell type annotation (LLMs frozen, only small MLP trained)                                               |
| ChatCell       | üìù‚ùå[Fang et al. 2024](https://arxiv.org/abs/2402.08303)                                                                                                                   | [üõ†GitHub](https://github.com/zjunlp/ChatCell)              | scRNA-seq                          | Natural Language [T5](https://huggingface.co/docs/transformers/en/model_doc/t5) and [natural language instructions](https://huggingface.co/datasets/zjunlp/ChatCell-Instructions)                                                                                                                                                                          | Other: Ordering with embedding as natural language with additional terms                                                         | Encoder-Decoder, [T5](https://huggingface.co/docs/transformers/en/model_doc/t5)                                     | NTP with CE loss                                                    | None (conditional sequence generation, prompting)                      | Simulation, cell type annotation, drug sensitivity prediction                                            |
| MarkerGeneBERT | üìù[Cheng et al. 2023](https://www.biorxiv.org/content/10.1101/2024.01.30.578115v1)                                                                                        | None                                                        | scRNA-seq                          | Natural Language, [PubMed](https://pubmed.ncbi.nlm.nih.gov/) and [PubMed Central](https://www.ncbi.nlm.nih.gov/pmc/)                                                                                                                                                                                                                                       | Other: Natural language preprocessed with [SciBERT](https://arxiv.org/abs/1903.10676)                                            | Encoder                                                                                                             | MLM                                                                 | Named Entity Recognition (NER), cell-biomarker sentence classification | None                                                                                                     |
| scELMo         | üìù[Liu, Chen and Zheng 2023](https://www.biorxiv.org/content/10.1101/2023.12.07.569910v1.full.pdf)                                                                        | [Partial üîçGitHub](https://github.com/HelloWorldLTY/scELMo) | scRNA-seq, CITE-seq                | Natural Language, Closed source                                                                                                                                                                                                                                                                                                                            | Other: NLP model embeddings of features weighted by the feature level in a cell (e.g. expression level)                          | Closed source (some open)                                                                                           | Closed source (some open)                                           | Cell type annotation, Genetic perturbation effect prediction           | Cell and gene embeddings in other perturbation models                                                    |
| GenePT         | üìù[Chen and Zou 2023](https://www.biorxiv.org/content/10.1101/2023.10.16.562533v1.full)                                                                                   | [Partial üîçGitHub](https://github.com/yiqunchen/GenePT)     | scRNA-seq                          | Natural Language, Closed source                                                                                                                                                                                                                                                                                                                            | Ordering: embedding as natural language                                                                                          | Closed source                                                                                                       | Closed source                                                       | Gene function prediction                                               | Cell clustering, GRN inference                                                                           |
| GPT-4          | üìÑ(Nature Meth)[W. Hou and Z. Ji 2024](https://www.nature.com/articles/s41592-024-02235-4)                                                                                | [üîçGitHub](https://github.com/Winnie09/GPTCelltype)         | scRNA-seq                          | Natural Language, Closed source                                                                                                                                                                                                                                                                                                                            | Ordering: embedding as natural language                                                                                          | Closed source                                                                                                       | Closed source                                                       | None (conditional sequence generation, prompting)                      | Cell type annotation                                                                                     |
| Cell2Sentence  | üìÑ(ICML)[Levine et al. 2024](https://openreview.net/forum?id=EWt5wsEdvc&referrer=%5Bthe%20profile%20of%20Josue%20Ortega%20Caro%5D(%2Fprofile%3Fid%3D~Josue_Ortega_Caro1)) | [üõ†Ô∏èGitHub](https://github.com/vandijklab/cell2sentence-ft) | scRNA-seq                          | Natural Language ([GPT2](https://d4mucfpksywv.cloudfront.net/better-language-models/language_models_are_unsupervised_multitask_learners.pdf)) and [scRNA-seq](https://www.science.org/doi/full/10.1126/science.abl5197?casa_token=KSZInYXxqU4AAAAA%3AuNgeqoX4vxOaMPGAv4UW9_GMy1lMmZ1-QGyx2VBCSbsGWvchKCzdNUvwt-h_yemzugH075TGz6N8fw) (40k / immune, human) | Ordering: embedding as natural language                                                                                          | Decoder                                                                                                             | NTP with CE loss                                                    | None                                                                   | Simulation, cell type annotation                                                                         | 


## Single-cell transformer evaluation

| Paper                                                                                    | Code                                                                                         | Omic Modalities                                | Evaluated Transformers                  | Tasks                                                                                                                                                                                | Notes                                                                                                                                                                                                                                                                                                                                   |
|------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------|------------------------------------------------|-----------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| üìù[Ahlmann-Eltze et al. 2024](https://www.biorxiv.org/content/10.1101/2024.09.16.613342) | [üõ†Ô∏èGitHub](https://github.com/const-ae/linear_perturbation_prediction-Paper/)                | scRNA-seq                                      | scGPT, scFoundation                     | Genetic perturbation effect prediction                                                                                                                                               | A simple linear model performs better than scGPT and scFoundation (and GEARS).                                                                                                                                                                                                                                                          |
| üìù[He et al. 2024](https://www.biorxiv.org/content/10.1101/2024.01.27.577455v1)          | [üõ†Ô∏èGitHub](https://github.com/laolintou/scPEFT)                                              | scRNA-seq                                      | scGPT                                   | Cell type annotation                                                                                                                                                                 | Evaluation of Parameter-Efficient Fine-Tuning (PEFT) for scGPT. Indicates that PEFT not only is more compute-efficient, but also results in better cell type prediction.                                                                                                                                                                |
| üìÑ(Nature MI)[Khan et al. 2023](https://www.nature.com/articles/s42256-023-00757-8)      | [üõ†Ô∏èGitHub](https://github.com/TranslationalBioinformaticsUnit/scbert-reusability)            | scRNA-seq                                      | scBERT                                  | Cell type annotation. Unseen cell type detection                                                                                                                                     | Focused on imbalanced cell type classification. scBERT is sensitive to class imbalance. scBERT outperforms Seurat. scBERT doesn't perform well in unseen cell type detection. It benefits from SSL pretraining.                                                                                                                         |
| üìù[Liu et al. 2023](https://www.biorxiv.org/content/10.1101/2023.09.08.555192v4)         | [üõ†Ô∏èGitHub](https://github.com/HelloWorldLTY/scEval)                                          | scRNA-seq, scATAC-seq, Spatial transcriptomics | scGPT, Geneformer, scBERT, tGPT, CellLM | Cell clustering, cell type annotation, multimodal embedding, GRN inference, gene expression imputation, genetic perturbation effect prediction, simulation, gene function prediction | Models aren't trained on the same datasets. scGPT is positioned as most versatile in terms of task diversity that it can tackle. Models other than transformer appear to be at least as good as transformers in most tasks. Transformers were shown to be sensitive to the choice of hyperparameters, such as learning rate and epochs. |
| üìù[Boiarsky et al. 2023](https://www.biorxiv.org/content/10.1101/2023.10.19.563100v1)    | [üõ†Ô∏èGitHub](https://github.com/clinicalml/sc-foundation-eval)                                 | scRNA-seq                                      | scBERT, scGPT                           | Cell type annotation                                                                                                                                                                 | Logistic regression appears to be as good as transformers in cell type annotation, even in low-data scenarios.                                                                                                                                                                                                                          |
| üìù[Kedzierska et al. 2023](https://www.biorxiv.org/content/10.1101/2023.10.16.561085v2)  | [üõ†Ô∏èGitHub](https://github.com/microsoft/zero-shot-scfoundation)                              | scRNA-seq                                      | scGPT, Geneformer                       | Cell clustering                                                                                                                                                                      | Zero-shot performance only. Both models appear unreliable.                                                                                                                                                                                                                                                                              |
| üìù[Alsabbagh et al. 2023](https://www.biorxiv.org/content/10.1101/2023.10.24.563625v1)   | [üõ†Ô∏èGitHub](https://github.com/SabbaghCodes/ImbalancedLearningForSingleCellFoundationModels)  | scRNA-seq                                      | scGPT, Geneformer, scBERT               | Cell type annotation                                                                                                                                                                 | Focused on imbalanced cell type classification. Geneformer appears to be outperformed by scGPT and scBERT, where the two latter perform similarly.                                                                                                                                                                                      |

#### Legend
- üìù - Preprint
- üìÑ - Peer-Reviewed Publication
- üõ†Ô∏è - Fully reproducible
- üîç - Code for evaluation only
- ‚ùå - Retracted or withdrawn

### Citing this work
If you find the the data in this repository useful for your work, please cite:
```
@Article{szalata_transformers_2024,
	title = {Transformers in single-cell omics: a review and new perspectives},
	volume = {21},
	issn = {1548-7105},
	url = {https://doi.org/10.1038/s41592-024-02353-z},
	doi = {10.1038/s41592-024-02353-z},
	abstract = {Recent efforts to construct reference maps of cellular phenotypes have expanded the volume and diversity of single-cell omics data, providing an unprecedented resource for studying cell properties. Despite the availability of rich datasets and their continued growth, current single-cell models are unable to fully capitalize on the information they contain. Transformers have become the architecture of choice for foundation models in other domains owing to their ability to generalize to heterogeneous, large-scale datasets. Thus, the question arises of whether transformers could set off a similar shift in the field of single-cell modeling. Here we first describe the transformer architecture and its single-cell adaptations and then present a comprehensive review of the existing applications of transformers in single-cell analysis and critically discuss their future potential for single-cell biology. By studying limitations and technical challenges, we aim to provide a structured outlook for future research directions at the intersection of machine learning and single-cell biology.},
	pages = {1430--1443},
	number = {8},
	journaltitle = {Nature Methods},
	shortjournal = {Nature Methods},
	author = {Sza≈Çata, Artur and Hrovatin, Karin and Becker, S√∂ren and Tejada-Lapuerta, Alejandro and Cui, Haotian and Wang, Bo and Theis, Fabian J.},
	date = {2024-08-01},}
```
